src/main/java/dev/jbang/util/VersionChecker.java

```diff
@@ -0,0 +1,124 @@
+package dev.jbang.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+
+import dev.jbang.Settings;
+
+public class VersionChecker {
+	private static final String jbangVersionUrl = "https://github.com/jbangdev/jbang/releases/latest/download/version.txt";
+
+	private static final long DELAY_DAYS = 3;
+
+	/**
+	 * First we determine when was the last time we checked Jbang's latest version
+	 * and if enough time has passed (so as not to annoy the user with too frequent
+	 * messages) we perform the check and tell the user if a new version is
+	 * available. NB: The check is never performed if Quiet mode is enabled.
+	 */
+	public static void possiblyCheck() {
+		if (!Util.isQuiet() && shouldCheck()) {
+			check();
+		}
+	}
+
+	// Figure out if we should check now or wait for a couple of days
+	public static boolean shouldCheck() {
+		try {
+			Path versionFile = Settings.getCacheDir().resolve("version.txt");
+			if (Files.isRegularFile(versionFile)) {
+				FileTime ts = Files.getLastModifiedTime(versionFile);
+				long diff = Duration.between(ts.toInstant(), Instant.now()).toDays();
+				return diff > DELAY_DAYS;
+			}
+		} catch (IOException e) {
+			// ignore
+		}
+		return true;
+	}
+
+	/**
+	 * Checks if we're running the latest version of Jbang by comparing its version
+	 * against the latest version on GitHub. If there is a newer version we tell the
+	 * user about it and give them instructions on how to update.
+	 */
+	public static void check() {
+		String latestVersion = retrieveLatestVersion();
+		if (latestVersion != null) {
+			if (compareVersions(latestVersion, Util.getJbangVersion()) > 0) {
+				Util.infoMsg("There is a new version of jbang available!");
+				Util.infoMsg("You have version " + Util.getJbangVersion()
+						+ " and " + latestVersion + " is the latest.");
+				if (runningManagedJbang()) {
+					Util.infoMsg("Run 'jbang app install --force jbang' to update to the latest version.");
+				} else {
+					Util.infoMsg("Use your package manager to update jbang to the latest version,");
+					Util.infoMsg("or visit https://jbang.dev to download and install it yourself.");
+				}
+			} else {
+				Util.verboseMsg("jbang is up-to-date.");
+			}
+		}
+	}
+
+	// Determines and returns the latest Jbang version from GitHub
+	private static String retrieveLatestVersion() {
+		try {
+			Path versionFile = Util.downloadFile(jbangVersionUrl, Settings.getCacheDir().toFile());
```

Comment:
what happens if no network available or url slow to respond ?

Reply:
Not having a network is unnoticeable. A slow network would make Jbang seem to hang (but so would anything that tries to access the network).

Two options I can think of:

 - We could either do something in a separate thread, but I'm not sure I like that option because then the output could appear in unexpected places. (We could wait and always show the text at the end, but personally that feels wrong, if we show it, it should appear first)
 - Or we could perform the check in a place where we already know we're doing network access. For example in the `downloadAndCacheFile()` method. Then if it's slow then it's just slow for everything, one simple query isn't going to majorly change that. It does mean that if you never run anything that accesses the network then you'll never be notified of newer versions.

Reply:
Third option:

 - Keep it like it is now, but make sure the timeout is set to 2s or so.

Thought of this option because option 2 above is kind of intrusive. We have this nice clean code for downloading stuff and suddenly there's this version check in there. Just feels ugly.